src/Symfony/Component/Serializer/Context/Normalizer/AbstractNormalizerContextBuilder.php

```diff
@@ -0,0 +1,168 @@
+
+use Symfony\Component\Serializer\Context\ContextBuilder;
+use Symfony\Component\Serializer\Exception\InvalidArgumentException;
+use Symfony\Component\Serializer\Normalizer\AbstractNormalizer;
+
+/**
+ * A helper providing autocompletion for available AbstractNormalizer options.
+ *
+ * @author Mathias Arlaud <mathias.arlaud@gmail.com>
+ */
+abstract class AbstractNormalizerContextBuilder extends ContextBuilder
+{
+    /**
+     * Configures how many loops of circular reference to allow while normalizing.
+     *
+     * The value 1 means that when we encounter the same object a
+     * second time, we consider that a circular reference.
+     *
+     * You can raise this value for special cases, e.g. in combination with the
+     * max depth setting of the object normalizer.
+     *
+     * Must be strictly positive.
+     *
+     * @throws InvalidArgumentException
+     */
+    public function withCircularReferenceLimit(?int $circularReferenceLimit): static
+    {
+        if (null !== $circularReferenceLimit && $circularReferenceLimit <= 0) {
+            throw new InvalidArgumentException(sprintf('The circular reference limit must be strictly positive, "%d" given.', $circularReferenceLimit));
+        }
+
+        return $this->with(AbstractNormalizer::CIRCULAR_REFERENCE_LIMIT, $circularReferenceLimit);
+    }
+
+    /**
+     * Configures an object to be updated instead of creating a new instance.
+     *
+     * If you have a nested structure, child objects will be overwritten with
+     * new instances unless you set AbstractObjectNormalizer::DEEP_OBJECT_TO_POPULATE to true.
+     */
+    public function withObjectToPopulate(?object $objectToPopulate): static
+    {
+        return $this->with(AbstractNormalizer::OBJECT_TO_POPULATE, $objectToPopulate);
+    }
+
+    /**
+     * Configures groups containing attributes to (de)normalize.
+     *
+     * Eg: ['group1', 'group2']
+     *
+     * @param list<string>|string|null $groups
+     */
+    public function withGroups(array|string|null $groups): static
+    {
+        if (null === $groups) {
+            return $this->with(AbstractNormalizer::GROUPS, null);
+        }
+
+        return $this->with(AbstractNormalizer::GROUPS, (array) $groups);
+    }
+
+    /**
+     * Configures attributes to (de)normalize.
+     *
+     * For nested structures, this list needs to reflect the object tree.
+     *
+     * Eg: ['foo', 'bar', 'object' => ['baz']]
+     *
+     * @param array<string|array>|null $attributes
+     *
+     * @throws InvalidArgumentException
+     */
+    public function withAttributes(?array $attributes): static
+    {
+        $it = new \RecursiveIteratorIterator(new \RecursiveArrayIterator($attributes ?? []), \RecursiveIteratorIterator::LEAVES_ONLY);
+
+        foreach ($it as $attribute) {
+            if (!\is_string($attribute)) {
+                throw new InvalidArgumentException(sprintf('The attributes must be strings, "%s" given.', get_debug_type($attribute)));
```

Comment:
attribute**~~s~~** must be string**~~s~~**?

Reply:
As it's an array of attributes, WDYT about `Each attribute must be a string, "%s" given.`?

Reply:
seems good